### PR TITLE
Dev ublox

### PR DESCRIPTION
**Pull request by rickaha**
RTCM and heartbeat

### CODE REVIEW

File: gnss/ubloxrover.cpp [hunk 1]
```diff
@@ -158,3 +180,16 @@ void UbloxRover::updateAHRS(const ubx_esf_meas &meas)
...
+void UbloxRover::updateGNSS(const ubx_nav_pvt &pvt)
+{
+    //qDebug() << "Current enuRef:" << mEnuReference.latitude << mEnuReference.longitude << mEnuReference.height;
+    llh_t llh = {pvt.lat, pvt.lon, pvt.height};
+    xyz_t xyz = coordinateTransforms::llhToEnu(mEnuReference, llh);
+    PosPoint gnssPos = mVehicleState->getPosition(PosType::GNSS);
+    gnssPos.setX(xyz.x);
+    gnssPos.setY(xyz.y);
+    gnssPos.setHeight(xyz.z);
+
```
Comment (marvind) [inline gnss/ubloxrover.cpp:193]: we should add orientation (if data is valid)

File: legacy/packetinterfacetcpserver.cpp [hunk 2]
```diff
@@ -19,167 +24,199 @@ PacketInterfaceTCPServer::PacketInterfaceTCPServer(QObject *parent) : QObject(pa
...
+                    ret.vbAppendDouble32(wpFollowerState.purePursuitRadius, 1e6); // autopilot_pp_radius
+                    ret.vbAppendInt32(QTime::currentTime().addSecs(-QDateTime::currentDateTime().offsetFromUtc()).msecsSinceStartOfDay());
+                    ret.vbAppendInt16(0); // autopilot_route_left
+                    ret.vbAppendDouble32(mVehicleState->getPosition(PosType::UWB).getX(), 1e4); // UWB px
+                    ret.vbAppendDouble32(mVehicleState->getPosition(PosType::UWB).getY(), 1e4); // UWB PY
+                    mTcpServer.packet()->sendPacket(ret);
                 }
+            } break;
+                // --- Set state on vehicle
+            case CMD_SET_POS:
```
Comment (marvind) [inline legacy/packetinterfacetcpserver.cpp:96]: this needs to go directly above CMD_SET_POS_ACK and the case needs to be changed like it was before: CMD_SET_POS and CMD_SET_POS_ACK do exactly the same, if it was actually CMD_SET_POS_ACK then ack is sent back

**Review by marvind — CHANGES_REQUESTED** [review]
CMD_SET_POS / CMD_SET_POS_ACK need to be changed to how they looked like before. Looks good otherwise!

File: legacy/packetinterfacetcpserver.cpp (hunk 2 above)
Comment (rickaha) [reply legacy/packetinterfacetcpserver.cpp:96]: Why do we want to use case for acks here? The acks are sent back to base station and are handled there. Before it was like this:

        // --- Set state on vehicle
        case CMD_SET_POS:
        case CMD_SET_POS_ACK: {
            PosPoint tmpPos = mVehicleState->getPosition();
            tmpPos.setX(packetData.vbPopFrontDouble32(1e4));
            tmpPos.setY(packetData.vbPopFrontDouble32(1e4));
            tmpPos.setYaw(packetData.vbPopFrontDouble32(1e6) * M_PI/180.0);
            mVehicleState->setPosition(tmpPos);

            if (commandID == CMD_SET_POS_ACK) {
                VByteArray ack;
                ack.vbAppendUint8(mVehicleState->getId());
                ack.vbAppendUint8(commandID);
                mTcpServer.packet()->sendPacket(ack);
            }
        } break;
        // --- Remote control vehicle

**Review by marvind — APPROVED** [review]
Fin!

**Pull request merged**